src/groupby/hash/groupby_compute_api.h

```diff
@@ -0,0 +1,210 @@
+
+  //the_map->prefetch(0);
+
+  CUDA_RT_CALL(cudaGetLastError());
+
+  // Inserts (groupby_column[i], aggregation_column[i]) as a key-value pair into the
+  // hash table. When a given key already exists in the table, the aggregation operation
+  // is computed between the new and existing value, and the result is stored back.
+  build_aggregation_table<<<build_grid_size, block_size>>>(the_map.get(), 
+                                                           in_groupby_column, 
+                                                           in_aggregation_column,
+                                                           in_column_size,
+                                                           aggregation_op);
+  CUDA_RT_CALL(cudaGetLastError());
+
+  // Used by threads to coordinate where to write their results
+  unsigned int * global_write_index{nullptr};
+  CUDA_RT_CALL(cudaMallocManaged(&global_write_index, sizeof(unsigned int)));
+  CUDA_RT_CALL(cudaMemset(global_write_index, 0, sizeof(unsigned int)));
+
+  CUDA_RT_CALL(cudaDeviceSynchronize());
+
+  const dim3 extract_grid_size ((the_map->size() + THREAD_BLOCK_SIZE - 1) / THREAD_BLOCK_SIZE, 1, 1);
+
+  // Extracts every non-empty key and value into separate contiguous arrays,
+  // which provides the result of the groupby operation
+  //extract_groupby_result<<<extract_grid_size, block_size>>>(the_map.get(),
+  //                                                          the_map->size(),
+  //                                                          out_groupby_column,
+  //                                                          out_aggregation_column,
+  //                                                          global_write_index);
+  
+  // FIXME Work around for above kernel failing to launch for some instantiations of the_map template class
+  map_type * map = the_map.get();
+  typename map_type::size_type map_size = the_map->size();
+  void *args[] = { &map, &map_size, &out_groupby_column, &out_aggregation_column, &global_write_index};
+
+  void (*func)(const map_type * const, 
+               const typename map_type::size_type, 
+               typename map_type::key_type * const, 
+               typename map_type::mapped_type * const,
+               unsigned int * const ) = &(extract_groupby_result<map_type>);
+
+  CUDA_RT_CALL(cudaLaunchKernel((const void*) func, extract_grid_size, block_size, args, 0, 0));
+  // FIXME End work around
+
+  CUDA_RT_CALL(cudaDeviceSynchronize());
+
+  // At the end of the extraction kernel, the global write index will be equal to
+  // the size of the output. Update the output size.
+  *out_size = *global_write_index;
+  CUDA_RT_CALL(cudaFree(global_write_index));
+
+  // Optionally sort the groupby/aggregation result columns
+  if(true == sort_result)
+  {
+    // Allocate double buffers needed for the cub Radix Sort
+    groupby_type * groupby_result_alt;
+    CUDA_RT_CALL(cudaMalloc(&groupby_result_alt, *out_size * sizeof(groupby_type)));
```

Comment:
Like my previous comments shouldn't we be moving towards the pattern that these compute functions don't do allocations? Or are we deciding that we will be doing allocations inside of them and that we will be relying on the some day to exist new memory manager to make them all fast?

Reply:
I think ideally compute functions wouldn't do allocations, but its a bit unavoidable in this instance. The caller of the groupby function won't know the size required for the temporary work buffers required for the CUB sort.

Furthermore, the sorting of the groupby result is only a stopgap until multi-column, hash-based groupby is supported. So this particular section of code will likely go away in the future.

Reply:
You know the worst case situation right? Do you think that the worse case is to conservative for the temp space?